pkg/guestagent/kubernetesservice/kubernetesservice.go

```diff
@@ -0,0 +1,130 @@
+package kubernetesservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type Protocol string
+
+const (
+	// UDP/SCTP when lima port forwarding works on those protocols
+	TCP Protocol = "TCP"
+)
+
+type Entry struct {
+	Protocol Protocol
+	IP       net.IP
+	Port     uint16
+}
+
+type ServiceWatcher struct {
+	serviceInformer cache.SharedIndexInformer
+}
+
+func NewServiceWatcher() *ServiceWatcher {
+	return &ServiceWatcher{}
+}
+
+func (s *ServiceWatcher) Start() {
+	retryInterval := 10 * time.Second
+	wait.PollInfinite(retryInterval, func() (done bool, err error) {
+		kubeClient, err := tryGetKubeClient()
+		if err != nil {
+			logrus.Infof("failed to get kube client: %v, will retry in %v", err, retryInterval)
+			return false, nil
+		}
+
+		ctx := context.TODO()
+		informerFactory := informers.NewSharedInformerFactory(kubeClient, time.Hour)
+		serviceInformer := informerFactory.Core().V1().Services().Informer()
+		informerFactory.Start(ctx.Done())
+		cache.WaitForCacheSync(ctx.Done(), serviceInformer.HasSynced)
+
+		s.serviceInformer = serviceInformer
+		return true, nil
+	})
+}
+
+func tryGetKubeClient() (kubernetes.Interface, error) {
+	candidateKubeConfigs := []string{
+		"/etc/rancher/k3s/k3s.yaml",
+		"/root/.kube/config",
```

Comment:
A more reliable way is to add kubernetes related config into lima YAML, but I feel it a bit exaggerated.